File: frontend/src/family/inheritance.test.ts
import { describe, expect, it } from 'vitest'
import type { Call, Person } from '../types'
import { alleleOrigins, layoutPedigree } from './inheritance'

const c = (g: string): Call => ({ rsid: 'rs1', chromosome: '1', position: 1, a1: g[0], a2: g[1] })
const from = (child: string, mum?: string, dad?: string) =>
  alleleOrigins(c(child), [
    { id: 'mum', call: mum ? c(mum) : undefined },
    { id: 'dad', call: dad ? c(dad) : undefined },
  ]).map((o) => o.from)

describe('alleleOrigins', () => {
  it('phases a heterozygous child from two homozygous parents', () => {
    expect(from('AG', 'AA', 'GG')).toEqual(['mum', 'dad'])
    expect(from('GA', 'AA', 'GG')).toEqual(['dad', 'mum'])
  })
  it('phases when only one assignment fits', () => {
    expect(from('AG', 'AG', 'GG')).toEqual(['mum', 'dad']) // dad cannot give A
  })
  it('attributes homozygous children to both parents', () => {
    expect(from('AA', 'AG', 'AG')).toEqual(['mum', 'dad'])
  })
  it('is ambiguous when both parents are heterozygous', () => {
    expect(from('AG', 'AG', 'AG')).toEqual(['ambiguous', 'ambiguous'])
  })
  it('flags impossible pairs', () => {
    expect(from('AA', 'GG', 'AG')).toEqual(['impossible', 'impossible'])
    expect(from('TT', 'CC')).toEqual(['impossible', 'impossible'])
  })
  it('handles one typed parent', () => {
    expect(from('AG', 'AA')).toEqual(['mum', 'dad']) // G must be dad's, though untyped
    expect(from('AG', undefined, 'GG')).toEqual(['mum', 'dad'])
    expect(from('AG', 'AG')).toEqual(['ambiguous', 'ambiguous'])
    expect(from('AA', 'AG')).toEqual(['mum', 'dad'])
  })
  it('returns nothing for no-calls and marks untyped parents', () => {
    expect(from('A-', 'AA', 'GG')).toEqual([])
    expect(from('AG')).toEqual(['untyped', 'untyped'])
  })
})

describe('layoutPedigree', () => {
  const p = (id: string, createdAt: string): Person => ({
    id,
    label: id,
    displayName: id,
    sex: 'unknown',
    birthYear: null,
    notes: '',
    createdAt,
  })
  it('puts founders on top and children under the mean of their parents', () => {
    const nodes = layoutPedigree(
      [p('kid', '3'), p('gran', '0'), p('mum', '1'), p('dad', '2'), p('uncle', '4')],
      [
        { parentId: 'gran', childId: 'mum' },
        { parentId: 'gran', childId: 'uncle' },
        { parentId: 'mum', childId: 'kid' },
        { parentId: 'dad', childId: 'kid' },
      ],
    )
    const at = (id: string) => nodes.find((n) => n.person.id === id)!
    expect([at('gran').generation, at('dad').generation, at('mum').generation, at('kid').generation]).toEqual(
      [0, 0, 1, 2],
    )
    expect(at('mum').column).toBeLessThan(at('uncle').column) // both under gran; creation order
    expect(at('kid').parents.sort()).toEqual(['dad', 'mum'])
  })
})

File: mobile/android/app/src/test/kotlin/com/hearth/family/InheritanceTest.kt
package com.hearth.family

import com.hearth.data.Call
import com.hearth.data.Person
import com.hearth.data.Relationship
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/** Mirrors frontend/src/family/inheritance.test.ts. */
class InheritanceTest {
    private fun c(g: String) = Call("rs1", "1", 1, g.substring(0, 1), g.substring(1, 2))
    private fun from(child: String, mum: String? = null, dad: String? = null) =
        alleleOrigins(c(child), listOf(ParentCall("mum", mum?.let(::c)), ParentCall("dad", dad?.let(::c)))).map { it.from }

    @Test fun phases() {
        assertEquals(listOf("mum", "dad"), from("AG", "AA", "GG"))
        assertEquals(listOf("dad", "mum"), from("GA", "AA", "GG"))
        assertEquals(listOf("mum", "dad"), from("AG", "AG", "GG"))
        assertEquals(listOf("mum", "dad"), from("AA", "AG", "AG"))
        assertEquals(listOf("ambiguous", "ambiguous"), from("AG", "AG", "AG"))
    }

    @Test fun `flags impossible pairs`() {
        assertEquals(listOf("impossible", "impossible"), from("AA", "GG", "AG"))
        assertEquals(listOf("impossible", "impossible"), from("TT", "CC"))
    }

    @Test fun `handles one typed parent`() {
        assertEquals(listOf("mum", "dad"), from("AG", "AA"))
        assertEquals(listOf("mum", "dad"), from("AG", null, "GG"))
        assertEquals(listOf("ambiguous", "ambiguous"), from("AG", "AG"))
        assertEquals(listOf("mum", "dad"), from("AA", "AG"))
    }

    @Test fun `no-calls and untyped parents`() {
        assertEquals(emptyList<String>(), from("A-", "AA", "GG"))
        assertEquals(listOf("untyped", "untyped"), from("AG"))
    }

    private fun p(id: String, createdAt: String) = Person(id, id, id, "unknown", null, "", createdAt)

    @Test fun `founders on top, co-parents side by side, children under them`() {
        val nodes = layoutPedigree(
            listOf(p("kid", "3"), p("gran", "0"), p("mum", "1"), p("dad", "2"), p("uncle", "4")),
            listOf(Relationship("gran", "mum"), Relationship("gran", "uncle"), Relationship("mum", "kid"), Relationship("dad", "kid")),
        )
        fun at(id: String) = nodes.first { it.person.id == id }
        assertEquals(listOf(0, 1, 1, 2), listOf(at("gran").generation, at("dad").generation, at("mum").generation, at("kid").generation))
        assertTrue(at("mum").column < at("uncle").column)
        assertEquals(1, Math.abs(at("dad").column - at("mum").column))
        assertEquals(listOf("dad", "mum"), at("kid").parents.sorted())
    }

    @Test fun `keeps a child below a parent pulled down to a co-parent`() {
        val nodes = layoutPedigree(
            listOf(p("gran", "0"), p("mum", "1"), p("dad", "2"), p("kid", "3"), p("step", "4"), p("half", "5")),
            listOf(
                Relationship("gran", "mum"), Relationship("mum", "kid"), Relationship("dad", "kid"),
                Relationship("dad", "half"), Relationship("step", "half"),
            ),
        )
        fun at(id: String) = nodes.first { it.person.id == id }
        assertEquals(listOf(1, 2), listOf(at("step").generation, at("half").generation))
    }
}
